### ISSUE TITLE
REST api function for retrieving all concepts present in a vocabulary

### ISSUE THREAD

**Issue by d0rg0ld** [NatLibFi/Skosmos#626]
Hi,

As far as I understand the current FINTO/SKOSMOS API doc, I can retrieve information about all the vocabularies present in the collection and can then retrieve top level concepts from or search for specific concepts within each of them. What I am missing is the possibility to directly harvest all declared concepts from each vocabulary, the only currently available option in this regard seems to be an API triggered RDF dump of the whole vocabulary. Probably I did not understand the API documentation well enough and it is already possible, in this case I'd be glad to learn how to do it, otherwise I'd like to suggest this as future enhancement.

Kind regards,
Doron Goldfarb

**Comment by osma** [comment]
You're right, getting a list of all the concepts from a vocabulary with the current API is quite awkward. It's not really intended for that purpose originally. As you mentioned, asking for the full vocabulary is probably the easiest way.

Can you elaborate on what kind of use case you have and what information you would need for each concept in the response?

**Comment by d0rg0ld** [comment]
Hi,

thanks for your reply! In the context of the EUDAT project we are currently working on a prototype for aggregating the contents from multiple semantic repositories, i.e. ontology hubs such as BioPortal etc., for providing centralized concept search and overlap analysis functionality - For more info I can offer a link to a recent workshop paper on our endeavor: http://ceur-ws.org/Vol-1933/paper-7.pdf.  

Coming from the environmental domain where we have a number of SKOS based resources, I'd really also like to include SKOS related repositories such as FINTO/SKOSMOS in our aggregator, because we consider re-using SKOSMOS for our own resources and it would be great if we could then directly harvest all stored concepts for aggregation. 

Regarding the information for each concept, the basics would include concept IRI, preferred label, synonyms and description, information we could use to provide lookup services to find the right concept for e.g. semantic annotation. 

I hope that this use case is of interest to you, I am looking forward to learning your opinion on this!

Thanks and kind regards,
Doron

**Comment by d0rg0ld** [comment]
Hi, 

I played around with the code a bit and extended the API with a new call "allConcepts". 

As far as my use case was concerned, it works fine, returning ID, prefLabel, altLabel(s) and Description for each concept present in a vocabulary. The only thing I did not add yet is paging, which might be necessary for larger vocabularies.

In case you are interested, these are the files were I made changes, you can see the diff in a Skosmos fork on https://github.com/d0rg0ld/Skosmos, I forked it today from the current version and merged my changes there.

https://github.com/d0rg0ld/Skosmos/blob/master/model/sparql/GenericSparql.php
https://github.com/d0rg0ld/Skosmos/blob/master/model/Vocabulary.php
https://github.com/d0rg0ld/Skosmos/blob/master/controller/RestController.php
https://github.com/d0rg0ld/Skosmos/blob/master/rest.php

Please let me know what you think about it...

Thanks and best regards,
Doron

**Comment by osma** [comment]
Hi @d0rg0ld , can you please make a pull request with your suggested changes? That would be much easier to review...

**Comment by d0rg0ld** [comment]
hi, thanks, Ill give it a try, haven't done that before

**Comment by d0rg0ld** [comment]
Hi, I just issued a pull request, thanks for having a look at my suggestions!
